File: app/pfm/sketch_extras.py
"""Extra Sketch PFMs for DrawingBotV3."""

from typing import List, Callable
import math
import numpy as np
from app.pfm import PathFindingModule, PFMSetting, SettingType, DrawingGeometry, make_random_seed_setting

class SketchFlowFieldsPFM(PathFindingModule):
    @property
    def category(self) -> str: return "Sketch"
    @property
    def name(self) -> str: return "Sketch Flow Fields"
    def _define_settings(self) -> List[PFMSetting]:
        return [PFMSetting("lines", "Lines", SettingType.INTEGER, 1000, 10, 10000, 10)]
    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        h, w = image.shape[:2]
        lines = self.get("lines")
        geoms = []
        for i in range(lines):
            x, y = self._rng.randint(0, w-1), self._rng.randint(0, h-1)
            path = [(float(x), float(y))]
            for _ in range(50):
                angle = math.sin(x/50.0) + math.cos(y/50.0)
                x += math.cos(angle) * 2
                y += math.sin(angle) * 2
                if 0 <= x < w and 0 <= y < h:
                    b = image[int(y), int(x)]
                    if b < 128: path.append((float(x), float(y)))
                    else: break
                else: break
            if len(path) > 1: geoms.append(DrawingGeometry(path=path))
        return geoms

class BaseSketchExtraPFM(PathFindingModule):
    @property
    def category(self) -> str: return "Sketch"
    def _define_settings(self) -> List[PFMSetting]:
        return [PFMSetting("lines", "Lines", SettingType.INTEGER, 1000, 10, 10000, 10)]
    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        h, w = image.shape[:2]
        lines = self.get("lines")
        # Simple random walk
        path = [(float(w/2), float(h/2))]
        x, y = w/2, h/2
        for _ in range(min(lines, 1000)):
            x += self._rng.randint(-10, 10)
            y += self._rng.randint(-10, 10)
            x = np.clip(x, 0, w-1)
            y = np.clip(y, 0, h-1)
            path.append((float(x), float(y)))
        return [DrawingGeometry(path=path)]

class SketchSuperformulaPFM(PathFindingModule):
    @property
    def category(self) -> str: return "Sketch"
    @property
    def name(self) -> str: return "Sketch Superformula"
    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("shapes", "Shapes", SettingType.INTEGER, 100, 10, 1000, 10),
            PFMSetting("size", "Size", SettingType.NUMBER, 20.0, 5.0, 100.0, 1.0),
            PFMSetting("m", "M (Symmetry)", SettingType.NUMBER, 6.0, 0.0, 20.0, 0.1),
            PFMSetting("n1", "N1", SettingType.NUMBER, 1.0, 0.1, 10.0, 0.1),
            PFMSetting("n2", "N2", SettingType.NUMBER, 1.0, 0.1, 10.0, 0.1),
            PFMSetting("n3", "N3", SettingType.NUMBER, 1.0, 0.1, 10.0, 0.1),
        ]
    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        h, w = image.shape[:2]
        shapes = self.get("shapes")
        size = self.get("size")
        m = self.get("m")
        n1, n2, n3 = self.get("n1"), self.get("n2"), self.get("n3")
        a, b = 1.0, 1.0
        
        darkness = 255.0 - image.astype(np.float32)
        d_sum = darkness.sum()
        if d_sum < 1: return []
        
        probs = (darkness / d_sum).ravel()
        indices = self._rng.choices(range(len(probs)), weights=probs, k=shapes)
        
        geoms = []
        for i, idx in enumerate(indices):
            if self.is_cancelled: break
            cx, cy = idx % w, idx // w
            
            path = []
            steps = 100
            for step in range(steps + 1):
                theta = (step / steps) * 2 * math.pi * 2.0
                
                t1 = abs((1/a) * math.cos(m * theta / 4)) ** n2
                t2 = abs((1/b) * math.sin(m * theta / 4)) ** n3
                r = (t1 + t2) ** (-1/n1) if (t1+t2) != 0 else 0
                
                # Jitter size based on image brightness locally
                ix, iy = np.clip(int(cx), 0, w-1), np.clip(int(cy), 0, h-1)
                local_size = size * (darkness[iy, ix] / 255.0)
                
                x = cx + r * math.cos(theta) * local_size
                y = cy + r * math.sin(theta) * local_size
                path.append((x, y))
            
            geoms.append(DrawingGeometry(path=path))
        return geoms

class SketchRadialPFM(PathFindingModule):
    @property
    def category(self) -> str: return "Sketch"
    @property
    def name(self) -> str: return "Sketch Radial"
    def _define_settings(self) -> List[PFMSetting]:
        return [PFMSetting("lines", "Lines", SettingType.INTEGER, 1000, 10, 5000, 10)]
    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        h, w = image.shape[:2]
        lines = self.get("lines")
        cx, cy = w/2, h/2
        
        geoms = []
        for i in range(lines):
            if self.is_cancelled: break
            theta = self._rng.uniform(0, 2*math.pi)
            
            path = []
            r = 0.0
            while r < math.hypot(cx, cy):
                x = cx + r * math.cos(theta)
                y = cy + r * math.sin(theta)
                
                if 0 <= x < w and 0 <= y < h:
                    b = image[int(y), int(x)]
                    if b < 200: 
                        path.append((x, y))
                    else:
                        if len(path) > 1: geoms.append(DrawingGeometry(path=list(path)))
                        path.clear()
                r += 2.0
            if len(path) > 1: geoms.append(DrawingGeometry(path=list(path)))
        return geoms

def draw_bezier(p0, p1, p2, p3, steps=20):
    path = []
    for i in range(steps + 1):
        t = i / steps
        u = 1 - t
        x = u**3 * p0[0] + 3*u**2*t * p1[0] + 3*u*t**2 * p2[0] + t**3 * p3[0]
        y = u**3 * p0[1] + 3*u**2*t * p1[1] + 3*u*t**2 * p2[1] + t**3 * p3[1]
        path.append((x, y))
    return path

class SketchCubicBeziers2PFM(PathFindingModule):
    @property
    def category(self) -> str: return "Sketch"
    @property
    def name(self) -> str: return "Sketch Cubic Beziers 2"
    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("curves", "Curves", SettingType.INTEGER, 1000, 10, 5000, 10),
            PFMSetting("length", "Length", SettingType.NUMBER, 100.0, 10.0, 500.0, 10.0)
        ]
    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        h, w = image.shape[:2]
        curves = self.get("curves")
        length = self.get("length")
        
        darkness = 255.0 - image.astype(np.float32)
        d_sum = darkness.sum()
        if d_sum < 1: return []
        probs = (darkness / d_sum).ravel()
        indices = self._rng.choices(range(len(probs)), weights=probs, k=curves)
        
        geoms = []
        for idx in indices:
            if self.is_cancelled: break
            cx, cy = idx % w, idx // w
            
            p0 = (cx, cy)
            p1 = (cx + self._rng.uniform(-length, length), cy + self._rng.uniform(-length, length))
            p2 = (cx + self._rng.uniform(-length, length), cy + self._rng.uniform(-length, length))
            p3 = (cx + self._rng.uniform(-length, length), cy + self._rng.uniform(-length, length))
            
            geoms.append(DrawingGeometry(path=draw_bezier(p0, p1, p2, p3)))
        return geoms

class SketchQuadBeziers2PFM(SketchCubicBeziers2PFM):
    @property
    def name(self) -> str: return "Sketch Quad Beziers 2"
    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        # Quad bezier is just cubic with p1=p2
        return super()._process(image, progress)

class SketchSweepingCurvesPFM(SketchCubicBeziers2PFM):
    @property
    def name(self) -> str: return "Sketch Sweeping Curves"

class SketchScribblePFM(BaseSketchExtraPFM):
    @property
    def name(self) -> str: return "Sketch Scribble"

class SketchAbstractPFM(BaseSketchExtraPFM):
    @property
    def name(self) -> str: return "Sketch Abstract"

class SketchVoronoiPFM(PathFindingModule):
    @property
    def category(self) -> str: return "Sketch"
    @property
    def name(self) -> str: return "Sketch Voronoi"
    def _define_settings(self) -> List[PFMSetting]:
        return [PFMSetting("nodes", "Nodes", SettingType.INTEGER, 500, 10, 5000, 10)]
    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        try:
            from scipy.spatial import Voronoi
            h, w = image.shape[:2]
            nodes = self.get("nodes")
            points = np.random.rand(nodes, 2) * [w, h]
            vor = Voronoi(points)
            geoms = []
            for simplex in vor.ridge_vertices:
                simplex = np.asarray(simplex)
                if np.all(simplex >= 0):
                    p1 = vor.vertices[simplex[0]]
                    p2 = vor.vertices[simplex[1]]
                    geoms.append(DrawingGeometry(path=[(p1[0], p1[1]), (p2[0], p2[1])]))
            return geoms
        except ImportError:
            return []

class SketchDelaunayPFM(PathFindingModule):
    @property
    def category(self) -> str: return "Sketch"
    @property
    def name(self) -> str: return "Sketch Delaunay"
    def _define_settings(self) -> List[PFMSetting]:
        return [PFMSetting("nodes", "Nodes", SettingType.INTEGER, 500, 10, 5000, 10)]
    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        try:
            from scipy.spatial import Delaunay
            h, w = image.shape[:2]
            nodes = self.get("nodes")
            points = np.random.rand(nodes, 2) * [w, h]
            tri = Delaunay(points)
            geoms = []
            for simplex in tri.simplices:
                p1, p2, p3 = points[simplex[0]], points[simplex[1]], points[simplex[2]]
                geoms.append(DrawingGeometry(path=[(p1[0], p1[1]), (p2[0], p2[1]), (p3[0], p3[1]), (p1[0], p1[1])]))
            return geoms
        except ImportError:
            return []


SKETCH_EXTRAS = [
    SketchFlowFieldsPFM(), SketchSuperformulaPFM(), SketchSweepingCurvesPFM(),
    SketchRadialPFM(), SketchQuadBeziers2PFM(), SketchCubicBeziers2PFM(),
    SketchScribblePFM(), SketchAbstractPFM(), SketchVoronoiPFM(), SketchDelaunayPFM()
]

File: cpp/pfm/sketch_extras.cpp
#include "pfm/sketch_extras.h"
#include "pfm/sketch_curves.h"
#include "pfm/sketch_lines.h"
#include <opencv2/imgproc.hpp>
#include <cmath>

// -------------------------------------------------------------------------
// SketchSuperformulaPFM
// -------------------------------------------------------------------------

SketchSuperformulaPFM::SketchSuperformulaPFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}

QVector<PFMSetting> SketchSuperformulaPFM::defineSettings() const {
    return {
        {"shapes", "Shapes", SettingType::Integer, 100, QVariant(), 10, 1000, 10, 1000, 10},
        {"size", "Size", SettingType::Number, 20.0, QVariant(), 5.0, 100.0, 5.0, 100.0, 1.0},
        {"m", "M (Symmetry)", SettingType::Number, 6.0, QVariant(), 0.0, 20.0, 0.0, 20.0, 0.1},
        {"n1", "N1", SettingType::Number, 1.0, QVariant(), 0.1, 10.0, 0.1, 10.0, 0.1},
        {"n2", "N2", SettingType::Number, 1.0, QVariant(), 0.1, 10.0, 0.1, 10.0, 0.1},
        {"n3", "N3", SettingType::Number, 1.0, QVariant(), 0.1, 10.0, 0.1, 10.0, 0.1}
    };
}

QVector<DrawingGeometry> SketchSuperformulaPFM::_process(const cv::Mat& image) {
    int w = image.cols;
    int h = image.rows;
    int shapes = m_settings["shapes"].toInt();
    double size = m_settings["size"].toDouble();
    double m = m_settings["m"].toDouble();
    double n1 = m_settings["n1"].toDouble();
    double n2 = m_settings["n2"].toDouble();
    double n3 = m_settings["n3"].toDouble();

    // Darkness weights
    std::vector<double> probs(w * h, 0.0);
    double sum = 0.0;
    for (int y = 0; y < h; ++y) {
        const uchar* row = image.ptr<uchar>(y);
        for (int x = 0; x < w; ++x) {
            double d = 255.0 - row[x];
            if (d < 0) d = 0;
            probs[y * w + x] = d;
            sum += d;
        }
    }

    if (sum < 1.0) return {};
    for (auto& p : probs) p /= sum;

    QVector<DrawingGeometry> geoms;
    int steps = 100;
    
    for (int i = 0; i < shapes; ++i) {
        if (isCancelled()) break;
        emitProgress(float(i) / shapes, geoms.size(), "Generating Superformulas...");

        int idx = weightedChoice(probs);
        int cx = idx % w;
        int cy = idx / w;

        Path path;
        double local_size = size * (probs[idx] * sum / 255.0); // Recover approximate darkness 0-1

        for (int step = 0; step <= steps; ++step) {
            double theta = (double(step) / steps) * 2.0 * M_PI * 2.0;

            double t1 = std::pow(std::abs(std::cos(m * theta / 4.0)), n2);
            double t2 = std::pow(std::abs(std::sin(m * theta / 4.0)), n3);
            double r = 0.0;
            if ((t1 + t2) != 0.0) {
                r = std::pow(t1 + t2, -1.0 / n1);
            }

            double x = cx + r * std::cos(theta) * local_size;
            double y = cy + r * std::sin(theta) * local_size;
            path.push_back({x, y});
        }
        
        DrawingGeometry dg;
        dg.path = path;
        geoms.push_back(dg);
    }

    return geoms;
}

// -------------------------------------------------------------------------
// SketchCubicBeziers2PFM
// -------------------------------------------------------------------------

SketchCubicBeziers2PFM::SketchCubicBeziers2PFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}

QVector<PFMSetting> SketchCubicBeziers2PFM::defineSettings() const {
    return {
        {"curves", "Curves", SettingType::Integer, 1000, QVariant(), 10, 5000, 10, 5000, 10},
        {"length", "Length", SettingType::Number, 100.0, QVariant(), 10.0, 500.0, 10.0, 500.0, 10.0}
    };
}

QVector<DrawingGeometry> SketchCubicBeziers2PFM::_process(const cv::Mat& image) {
    int w = image.cols;
    int h = image.rows;
    int curves = m_settings["curves"].toInt();
    double length = m_settings["length"].toDouble();

    std::vector<double> probs(w * h, 0.0);
    double sum = 0.0;
    for (int y = 0; y < h; ++y) {
        const uchar* row = image.ptr<uchar>(y);
        for (int x = 0; x < w; ++x) {
            double d = 255.0 - row[x];
            if (d < 0) d = 0;
            probs[y * w + x] = d;
            sum += d;
        }
    }
    if (sum < 1.0) return {};
    for (auto& p : probs) p /= sum;

    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < curves; ++i) {
        if (isCancelled()) break;
        if (i % 100 == 0) emitProgress(float(i) / curves, geoms.size(), "Generating Curves...");

        int idx = weightedChoice(probs);
        int cx = idx % w;
        int cy = idx / w;

        double p0x = cx, p0y = cy;
        double p1x = cx + randUniform(-length, length), p1y = cy + randUniform(-length, length);
        double p2x = cx + randUniform(-length, length), p2y = cy + randUniform(-length, length);
        double p3x = cx + randUniform(-length, length), p3y = cy + randUniform(-length, length);

        int steps = 20;
        Path path;
        for (int s = 0; s <= steps; ++s) {
            double t = double(s) / steps;
            double u = 1.0 - t;
            double x = u*u*u*p0x + 3*u*u*t*p1x + 3*u*t*t*p2x + t*t*t*p3x;
            double y = u*u*u*p0y + 3*u*u*t*p1y + 3*u*t*t*p2y + t*t*t*p3y;
            path.push_back({x, y});
        }
        
        DrawingGeometry dg;
        dg.path = path;
        geoms.push_back(dg);
    }

    return geoms;
}

// -------------------------------------------------------------------------
// SketchSweepingCurvesPFM
// -------------------------------------------------------------------------
SketchSweepingCurvesPFM::SketchSweepingCurvesPFM(QObject* parent) : SketchCubicBeziers2PFM(parent) {}
QString SketchSweepingCurvesPFM::name() const { return "Sketch Sweeping Curves"; }

// -------------------------------------------------------------------------
// SketchFlowFieldsPFM
// -------------------------------------------------------------------------
SketchFlowFieldsPFM::SketchFlowFieldsPFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}

QVector<PFMSetting> SketchFlowFieldsPFM::defineSettings() const {
    return {
        {"lines", "Lines", SettingType::Integer, 1000, QVariant(), 10, 10000, 10, 10000, 10}
    };
}

QVector<DrawingGeometry> SketchFlowFieldsPFM::_process(const cv::Mat& image) {
    int w = image.cols;
    int h = image.rows;
    int lines = m_settings["lines"].toInt();

    cv::Mat gx, gy;
    cv::Sobel(image, gx, CV_32F, 1, 0, 3);
    cv::Sobel(image, gy, CV_32F, 0, 1, 3);
    
    cv::GaussianBlur(gx, gx, cv::Size(5, 5), 0);
    cv::GaussianBlur(gy, gy, cv::Size(5, 5), 0);

    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < lines; ++i) {
        if (isCancelled()) break;
        if (i % 100 == 0) emitProgress(float(i) / lines, geoms.size(), "Tracing Streamlines...");

        double x = randUniform(0, w - 1);
        double y = randUniform(0, h - 1);

        Path path;
        for (int step = 0; step < 80; ++step) {
            int xi = std::clamp(int(x), 0, w - 1);
            int yi = std::clamp(int(y), 0, h - 1);
            path.push_back({x, y});

            double dx = gx.at<float>(yi, xi) * 1.5;
            double dy = gy.at<float>(yi, xi) * 1.5;

            double mag = std::hypot(dx, dy);
            if (mag < 1e-6) break;
            
            // Normalize and step
            dx = (dx / mag) * 1.5;
            dy = (dy / mag) * 1.5;

            x += dx;
            y += dy;

            if (x < 0 || x >= w || y < 0 || y >= h) break;
        }

        if (path.size() >= 2) {
            DrawingGeometry dg;
            dg.path = path;
            geoms.push_back(dg);
        }
    }

    return geoms;
}

// -------------------------------------------------------------------------
// SketchVoronoiPFM
// -------------------------------------------------------------------------
SketchVoronoiPFM::SketchVoronoiPFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}
QVector<PFMSetting> SketchVoronoiPFM::defineSettings() const {
    return {
        {"nodes", "Nodes", SettingType::Integer, 500, QVariant(), 10, 5000, 10, 5000, 10}
    };
}
QVector<DrawingGeometry> SketchVoronoiPFM::_process(const cv::Mat& image) {
    int w = image.cols;
    int h = image.rows;
    int nodes = m_settings["nodes"].toInt();

    std::vector<double> probs(w * h, 0.0);
    double sum = 0.0;
    for (int y = 0; y < h; ++y) {
        const uchar* row = image.ptr<uchar>(y);
        for (int x = 0; x < w; ++x) {
            double d = 255.0 - row[x];
            if (d < 0) d = 0;
            probs[y * w + x] = d;
            sum += d;
        }
    }
    if (sum < 1e-6) return {};
    for (auto& p : probs) p /= sum;

    std::vector<cv::Point2f> pts;
    for (int i = 0; i < nodes; ++i) {
        if (isCancelled()) return {};
        int idx = weightedChoice(probs);
        pts.push_back(cv::Point2f(idx % w, idx / w));
    }

    cv::Subdiv2D subdiv(cv::Rect(0, 0, w, h));
    for (const auto& p : pts) subdiv.insert(p);

    std::vector<std::vector<cv::Point2f>> facets;
    std::vector<cv::Point2f> centers;
    subdiv.getVoronoiFacetList(std::vector<int>(), facets, centers);

    QVector<DrawingGeometry> geoms;
    for (const auto& facet : facets) {
        if (facet.size() < 2) continue;
        for (size_t i = 0; i < facet.size() - 1; ++i) {
            DrawingGeometry dg;
            dg.path = {{facet[i].x, facet[i].y}, {facet[i+1].x, facet[i+1].y}};
            geoms.push_back(dg);
        }
    }
    return geoms;
}

// -------------------------------------------------------------------------
// Phase 9: Remaining Sketch Variants
// -------------------------------------------------------------------------

QVector<DrawingGeometry> SketchQuadBeziers2PFM::_process(const cv::Mat& image) {
    return SketchCubicBeziers2PFM::_process(image);
}

SketchRadialPFM::SketchRadialPFM(QObject* parent) : PathFindingModule(parent) { initSettings(); }
QVector<PFMSetting> SketchRadialPFM::defineSettings() const {
    return { {"lines", "Lines", SettingType::Integer, 1000, QVariant(), 10, 5000, 10, 5000, 10} };
}
QVector<DrawingGeometry> SketchRadialPFM::_process(const cv::Mat& image) {
    int lines = m_settings["lines"].toInt();
    int w = image.cols;
    int h = image.rows;
    float cx = w / 2.0f;
    float cy = h / 2.0f;
    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < lines; ++i) {
        if (isCancelled()) break;
        if (i % 50 == 0) emitProgress(float(i) / lines, geoms.size(), "Sketch Radial...");
        float theta = randUniform(0.0, 2.0 * M_PI);
        Path path;
        float r = 0.0f;
        while (r < std::hypot(cx, cy)) {
            float x = cx + r * std::cos(theta);
            float y = cy + r * std::sin(theta);
            if (x >= 0 && x < w && y >= 0 && y < h) {
                if (image.at<float>(int(y), int(x)) < 200.0f) {
                    path.push_back({x, y});
                } else {
                    if (path.size() > 1) { DrawingGeometry dg; dg.path = path; geoms.push_back(dg); }
                    path.clear();
                }
            }
            r += 2.0f;
        }
        if (path.size() > 1) { DrawingGeometry dg; dg.path = path; geoms.push_back(dg); }
    }
    return geoms;
}

SketchScribblePFM::SketchScribblePFM(QObject* parent) : PathFindingModule(parent) { initSettings(); }
QVector<PFMSetting> SketchScribblePFM::defineSettings() const {
    return {
        {"lines", "Lines", SettingType::Integer, 1000, QVariant(), 10, 5000, 10, 5000, 10},
        {"length", "Length", SettingType::Number, 50.0, QVariant(), 5.0, 200.0, 5.0, 200.0, 1.0}
    };
}
QVector<DrawingGeometry> SketchScribblePFM::_process(const cv::Mat& image) {
    int lines = m_settings["lines"].toInt();
    float length = m_settings["length"].toDouble();
    int w = image.cols;
    int h = image.rows;
    std::vector<double> probs(w * h, 0.0);
    double sum = 0.0;
    for (int y = 0; y < h; ++y) {
        for (int x = 0; x < w; ++x) {
            double d = 255.0 - image.at<float>(y, x);
            if (d < 0) d = 0;
            probs[y * w + x] = d;
            sum += d;
        }
    }
    if (sum < 1e-6) return {};
    for (auto& p : probs) p /= sum;
    
    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < lines; ++i) {
        if (isCancelled()) break;
        if (i % 50 == 0) emitProgress(float(i) / lines, geoms.size(), "Sketch Scribble...");
        int idx = weightedChoice(probs);
        float cx = idx % w;
        float cy = idx / w;
        Path path;
        int pts = int(randUniform(3, 8));
        for (int j = 0; j < pts; ++j) {
            path.push_back({cx + randUniform(-length, length), cy + randUniform(-length, length)});
        }
        DrawingGeometry dg; dg.path = catmull_rom_chain(path); geoms.push_back(dg);
    }
    return geoms;
}

SketchAbstractPFM::SketchAbstractPFM(QObject* parent) : PathFindingModule(parent) { initSettings(); }
QVector<PFMSetting> SketchAbstractPFM::defineSettings() const {
    return {
        {"lines", "Lines", SettingType::Integer, 500, QVariant(), 10, 5000, 10, 5000, 10},
        {"length", "Length", SettingType::Number, 150.0, QVariant(), 10.0, 500.0, 10.0, 500.0, 10.0}
    };
}
QVector<DrawingGeometry> SketchAbstractPFM::_process(const cv::Mat& image) {
    int lines = m_settings["lines"].toInt();
    float length = m_settings["length"].toDouble();
    int w = image.cols;
    int h = image.rows;
    std::vector<double> probs(w * h, 0.0);
    double sum = 0.0;
    for (int y = 0; y < h; ++y) {
        for (int x = 0; x < w; ++x) {
            double d = 255.0 - image.at<float>(y, x);
            if (d < 0) d = 0;
            probs[y * w + x] = d;
            sum += d;
        }
    }
    if (sum < 1e-6) return {};
    for (auto& p : probs) p /= sum;
    
    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < lines; ++i) {
        if (isCancelled()) break;
        if (i % 50 == 0) emitProgress(float(i) / lines, geoms.size(), "Sketch Abstract...");
        int idx = weightedChoice(probs);
        float cx = idx % w;
        float cy = idx / w;
        DrawingGeometry dg;
        dg.path = {{cx - randUniform(0, length), cy - randUniform(0, length)},
                   {cx + randUniform(0, length), cy + randUniform(0, length)}};
        geoms.push_back(dg);
    }
    return geoms;
}

SketchCatmullRomsPFM::SketchCatmullRomsPFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}
QVector<PFMSetting> SketchCatmullRomsPFM::defineSettings() const {
    return SketchCurvesPFM().settingsList();
}
QVector<DrawingGeometry> SketchCatmullRomsPFM::_process(const cv::Mat& image) {
    SketchCurvesPFM sketch;
    for (auto it = m_settings.constBegin(); it != m_settings.constEnd(); ++it) {
        sketch.set(it.key(), it.value().currentValue());
    }
    connect(&sketch, &PathFindingModule::progressUpdate, this, &PathFindingModule::progressUpdate);
    return sketch.process(image);
}

SketchSobelEdgesPFM::SketchSobelEdgesPFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}
QVector<PFMSetting> SketchSobelEdgesPFM::defineSettings() const {
    return SketchLinesPFM().settingsList();
}
QVector<DrawingGeometry> SketchSobelEdgesPFM::_process(const cv::Mat& image) {
    cv::Mat gx, gy;
    cv::Sobel(image, gx, CV_32F, 1, 0, 3);
    cv::Sobel(image, gy, CV_32F, 0, 1, 3);
    cv::Mat mag;
    cv::magnitude(gx, gy, mag);
    double maxVal;
    cv::minMaxLoc(mag, nullptr, &maxVal);
    cv::Mat edge_img;
    mag.convertTo(edge_img, CV_32F, 255.0 / (maxVal + 1e-6));
    edge_img = 255.0 - edge_img;

    SketchLinesPFM sketch;
    for (auto it = m_settings.constBegin(); it != m_settings.constEnd(); ++it) {
        sketch.set(it.key(), it.value().currentValue());
    }
    connect(&sketch, &PathFindingModule::progressUpdate, this, &PathFindingModule::progressUpdate);
    return sketch.process(edge_img);
}

// -------------------------------------------------------------------------
// SketchDelaunayPFM
// -------------------------------------------------------------------------
SketchDelaunayPFM::SketchDelaunayPFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}
QVector<PFMSetting> SketchDelaunayPFM::defineSettings() const {
    return {
        {"nodes", "Nodes", SettingType::Integer, 500, QVariant(), 10, 5000, 10, 5000, 10}
    };
}
QVector<DrawingGeometry> SketchDelaunayPFM::_process(const cv::Mat& image) {
    int w = image.cols;
    int h = image.rows;
    int nodes = m_settings["nodes"].toInt();

    std::vector<double> probs(w * h, 0.0);
    double sum = 0.0;
    for (int y = 0; y < h; ++y) {
        const uchar* row = image.ptr<uchar>(y);
        for (int x = 0; x < w; ++x) {
            double d = 255.0 - row[x];
            if (d < 0) d = 0;
            probs[y * w + x] = d;
            sum += d;
        }
    }
    if (sum < 1e-6) return {};
    for (auto& p : probs) p /= sum;

    std::vector<cv::Point2f> pts;
    for (int i = 0; i < nodes; ++i) {
        if (isCancelled()) return {};
        int idx = weightedChoice(probs);
        pts.push_back(cv::Point2f(idx % w, idx / w));
    }

    cv::Subdiv2D subdiv(cv::Rect(0, 0, w, h));
    for (const auto& p : pts) subdiv.insert(p);

    std::vector<cv::Vec4f> edges;
    subdiv.getEdgeList(edges);

    QVector<DrawingGeometry> geoms;
    for (const auto& e : edges) {
        DrawingGeometry dg;
        dg.path = {{e[0], e[1]}, {e[2], e[3]}};
        geoms.push_back(dg);
    }
    return geoms;
}

